File: bot.py
import asyncio
import logging
import time

import discord
from discord import app_commands
from discord.ext import commands

from core import config, db
from core.scheduling import get_scheduler
from web import server as web_server


settings = config.load()

# If we see this many gateway disconnects within the window, assume a doomed
# reconnect loop (e.g. session-invalidation cycle) and exit so we don't keep
# hammering Discord and deepening any rate-limit ban. Operator must restart
# manually after investigating.
DISCONNECT_THRESHOLD = 5
DISCONNECT_WINDOW_SECS = 300

logging.basicConfig(
    level=settings.log_level,
    format="%(asctime)s [%(levelname)s] %(name)s: %(message)s",
)
log = logging.getLogger("fo40")


# Add new cogs here as you build them.
INITIAL_COGS = [
    "cogs.scheduler",
    "cogs.mod_notes",
    # "cogs.dm_reports",
    # "cogs.reaction_roles",
    # "cogs.prompts",
    # "cogs.birthdays",
    "cogs.reddit_sync",   # self-disables if Reddit creds aren't set in .env
]


class FO40Bot(commands.Bot):
    def __init__(self):
        intents = discord.Intents.default()
        intents.members = True            # role checks, joins
        intents.message_content = False   # flip on if a future cog needs it
        # message_content is off, so prefix commands cannot read text. Use
        # mention-as-prefix as a no-op fallback; all real commands are slash.
        super().__init__(command_prefix=commands.when_mentioned, intents=intents)
        self._disconnect_times: list[float] = []
        self._web_runner = None

    async def setup_hook(self):
        await db.init_db()
        for cog in INITIAL_COGS:
            await self.load_extension(cog)
            log.info("Loaded %s", cog)
        self.tree.on_error = _on_app_command_error
        get_scheduler().start()
        self._web_runner = await web_server.start(self)
        # Sync slash commands to the guild for instant availability.
        # Commands default to global scope (~1hr Discord propagation), so copy
        # them into the guild's namespace first; then the guild sync pushes them
        # immediately.
        guild = discord.Object(id=settings.guild_id)
        self.tree.copy_global_to(guild=guild)
        await self.tree.sync(guild=guild)

    async def on_ready(self):
        log.info("Logged in as %s (%s)", self.user, self.user.id)

    async def on_resumed(self):
        log.info("gateway session resumed")

    async def on_disconnect(self):
        # Don't count disconnects during a planned shutdown.
        if self.is_closed():
            return
        now = time.monotonic()
        self._disconnect_times = [
            t for t in self._disconnect_times
            if now - t < DISCONNECT_WINDOW_SECS
        ]
        self._disconnect_times.append(now)
        log.warning(
            "gateway disconnect (%d in last %ds)",
            len(self._disconnect_times), DISCONNECT_WINDOW_SECS,
        )
        if len(self._disconnect_times) >= DISCONNECT_THRESHOLD:
            log.error(
                "%d disconnects in %ds — doomed reconnect loop suspected. "
                "Shutting down to prevent rate-limit damage. "
                "Wait at least 10 minutes and investigate before restarting.",
                len(self._disconnect_times), DISCONNECT_WINDOW_SECS,
            )
            await self.close()

    async def close(self):
        if self._web_runner is not None:
            try:
                await self._web_runner.cleanup()
            except Exception:
                log.exception("web server cleanup failed")
        sched = get_scheduler()
        if sched.running:
            sched.shutdown(wait=False)
        await super().close()


async def _on_app_command_error(
    interaction: discord.Interaction,
    error: app_commands.AppCommandError,
):
    """Send a clean ephemeral message for permission failures; log everything else."""
    if isinstance(error, app_commands.CheckFailure):
        msg = str(error) or "Permission denied."
        try:
            if interaction.response.is_done():
                await interaction.followup.send(msg, ephemeral=True)
            else:
                await interaction.response.send_message(msg, ephemeral=True)
        except discord.HTTPException:
            log.exception("failed to send check-failure response")
        return
    log.exception("app command error", exc_info=error)


async def main():
    bot = FO40Bot()
    async with bot:
        await bot.start(settings.discord_token)


if __name__ == "__main__":
    asyncio.run(main())

File: src/bot.ts
import {
  Client,
  Events,
  GatewayIntentBits,
  Interaction,
  REST,
  Routes,
  MessageFlags,
} from "discord.js";

import { loadSettings } from "./core/config.js";
import { getDb, closeDb } from "./core/db.js";
import { shutdownScheduler } from "./core/scheduling.js";
import { BotModule, ModuleCommand } from "./core/types.js";
import schedulerModule from "./modules/scheduler.js";
import modNotesModule from "./modules/modNotes.js";
import redditSyncModule from "./modules/redditSync.js";
import { startWebServer } from "./web/server.js";

const settings = loadSettings();

const MODULES: BotModule[] = [schedulerModule, modNotesModule, redditSyncModule];

// If we see this many gateway disconnects within the window, exit so we
// don't keep hammering Discord and deepen any rate-limit ban.
const DISCONNECT_THRESHOLD = 5;
const DISCONNECT_WINDOW_SECS = 300;

export async function run(): Promise<void> {
  // Initialise DB up front so schema migrations don't race with cogs.
  getDb();

  const client = new Client({
    intents: [
      GatewayIntentBits.Guilds,
      GatewayIntentBits.GuildMembers,
      GatewayIntentBits.GuildMessages,
      GatewayIntentBits.GuildInvites,
    ],
  });

  // Build a name -> handler map from each module's commands.
  const commandHandlers = new Map<string, ModuleCommand>();
  const commandJson: ModuleCommand["data"][] = [];
  for (const mod of MODULES) {
    for (const cmd of mod.commands ?? []) {
      commandHandlers.set(cmd.data.name, cmd);
      commandJson.push(cmd.data);
    }
  }

  client.on(Events.InteractionCreate, async (interaction: Interaction) => {
    if (!interaction.isChatInputCommand()) return;
    const handler = commandHandlers.get(interaction.commandName);
    if (!handler) {
      await interaction
        .reply({
          content: "Unknown command.",
          flags: MessageFlags.Ephemeral,
        })
        .catch(() => {});
      return;
    }
    try {
      await handler.execute(interaction);
    } catch (err) {
      console.error(
        `command ${interaction.commandName} threw:`,
        err,
      );
      try {
        if (interaction.replied || interaction.deferred) {
          await interaction.followUp({
            content: "Something went wrong handling that command.",
            flags: MessageFlags.Ephemeral,
          });
        } else {
          await interaction.reply({
            content: "Something went wrong handling that command.",
            flags: MessageFlags.Ephemeral,
          });
        }
      } catch {
        // ignore
      }
    }
  });

  // Disconnect-loop guard.
  const disconnectTimes: number[] = [];
  client.on(Events.ShardDisconnect, async (event) => {
    const now = Date.now() / 1000;
    while (disconnectTimes.length && now - disconnectTimes[0]! > DISCONNECT_WINDOW_SECS) {
      disconnectTimes.shift();
    }
    disconnectTimes.push(now);
    console.warn(
      `gateway disconnect (code=${event.code}; ${disconnectTimes.length} in last ${DISCONNECT_WINDOW_SECS}s)`,
    );
    if (disconnectTimes.length >= DISCONNECT_THRESHOLD) {
      console.error(
        `${disconnectTimes.length} disconnects in ${DISCONNECT_WINDOW_SECS}s — doomed reconnect loop suspected. ` +
          "Shutting down to prevent rate-limit damage. Wait 10+ min and investigate before restarting.",
      );
      await shutdown(client);
      process.exit(1);
    }
  });

  client.once(Events.ClientReady, async (c) => {
    console.info(`Logged in as ${c.user.tag} (${c.user.id})`);
  });

  // Initialise modules (registers listeners + scheduler jobs).
  for (const mod of MODULES) {
    if (mod.init) await mod.init(client);
    console.info(`Loaded module: ${mod.name}`);
  }

  // Start the web server (no-op if env vars not set).
  const fastify = await startWebServer(client);

  // Login to Discord.
  await client.login(settings.discordToken);

  // Sync slash commands to the guild for instant availability.
  if (commandJson.length > 0) {
    const rest = new REST({ version: "10" }).setToken(settings.discordToken);
    const appId = client.application?.id ?? client.user!.id;
    await rest.put(
      Routes.applicationGuildCommands(appId, settings.guildId),
      { body: commandJson },
    );
    console.info(`Synced ${commandJson.length} slash commands to guild`);
  }

  // Graceful shutdown on signal.
  const onSignal = async (sig: string) => {
    console.info(`received ${sig}; shutting down`);
    if (fastify) {
      try {
        await fastify.close();
      } catch (err) {
        console.error("fastify close failed:", err);
      }
    }
    await shutdown(client);
    process.exit(0);
  };
  process.on("SIGINT", () => void onSignal("SIGINT"));
  process.on("SIGTERM", () => void onSignal("SIGTERM"));
}

async function shutdown(client: Client): Promise<void> {
  shutdownScheduler();
  try {
    await client.destroy();
  } catch (err) {
    console.error("client destroy failed:", err);
  }
  closeDb();
}
